core/src/main/java/jenkins/console/ConsoleUrlProvider.java

```diff
@@ -0,0 +1,121 @@
+ * @since TODO
+ */
+public interface ConsoleUrlProvider extends Describable<ConsoleUrlProvider> {
+    @Restricted(NoExternalUse.class)
+    Logger LOGGER = Logger.getLogger(ConsoleUrlProvider.class.getName());
+
+    /**
+     * Get a URL relative to the context path of Jenkins which should be used to link to the console for the specified build.
+     * <p>Should only be used in the context of serving an HTTP request.
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the context of Jenkins, or {@code null}
+     * if this implementation does not want to server a special console view for this build.
+     */
+    @CheckForNull String getConsoleUrl(Run<?, ?> run);
+
+    @Override
+    default Descriptor<ConsoleUrlProvider> getDescriptor() {
+        return Jenkins.get().getDescriptorOrDie(getClass());
+    }
+
+    /**
+     * Get a URL relative to the web server root which should be used to link to the console for the specified build.
+     * <p>Should only be used in the context of serving an HTTP request.
+     * <p>Use {@link Functions#getConsoleUrl} to obtain this link in a Jelly template.
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the web server root
+     */
+    static @NonNull String getRedirectUrl(Run<?, ?> run) {
+        ConsoleUrlProviderGlobalConfiguration globalConfig = ConsoleUrlProviderGlobalConfiguration.get();
+        List<ConsoleUrlProvider> providers = globalConfig.getProviders();
+        User currentUser = User.current();
+        if (currentUser != null) { // TODO: Should we add a configuration option to allow admins to suppress this behavior?
+            ConsoleUrlProviderUserProperty userProperty = currentUser.getProperty(ConsoleUrlProviderUserProperty.class);
+            if (userProperty != null) {
+                List<ConsoleUrlProvider> userProviders = userProperty.getProviders();
+                if (userProviders != null) {
+                    providers = userProviders;
+                }
+            }
+        }
+        String url = null;
+        if (providers != null) {
+            for (ConsoleUrlProvider provider : providers) {
+                try {
+                    url = provider.getConsoleUrl(run);
+                    if (url != null) {
+                        break;
+                    }
+                } catch (Exception e) { // Intentionally broad catch clause to guard against broken implementations.
+                    LOGGER.log(Level.WARNING, e, () -> "Error looking up console URL for " + run + " from " + provider.getClass());
+                }
+            }
+        }
+        if (url == null) {
+            // Reachable if DefaultConsoleUrlProvider is not one of the configured providers, including if no providers are configured at all.
+            url = run.getUrl() + "console";
+        }
+        // TODO:
+        // * Fail if absolute URL?
```

Comment:
e.g., if an extension returns `https://www.example.com`, do we to support that case (skip context path stuff below), explicitly fail by throwing an exception, implicitly fail by building an invalid URL, etc.

Reply:
I'd treat it like a `null` URL returned from a provider, invalid return value.

Reply:
Done in [b5420ad](https://github.com/jenkinsci/jenkins/pull/8321/commits/b5420ad07c46402632bdd712a5cac7ffb88bda04). I also added a test for the new logic.